### PR TITLE
Social Media App milestone 2

### PR DESCRIPTION
**Pull request by Leon-Mbegera**
In this second milestone of the Social media app, a number of implementations come into play, i.e
- Only a logged-in user can see other users in the app and even see their posts and comments.
- A user can send an invitation to another user who is not yet a friend of theirs
- A user can receive an invitation from another user to become friends
- An option to either accept or decline friendship requests is also implemented
- Besides every user's name, it reads either 'pending or friends' in regards to the confirmation status of that friendship

### CODE REVIEW

**Review by mke2111 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

## Hey Happy Coder, 👋🏽 
**This is a great project you have going on, well-done lad**. 

### Well-done 💯 
    1. The app runs just fine

**However, there are a few sections to improve before your project is approved.** 🔴 
<hr>

### REQUIRED

- [x] Add a PR description explaining all the implemented features in this milestone.
- [x] The testing command in the readme file should be `rspec` and not `rpsec`.
- [x] Add unit and integration tests for your app. [Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png)
- [x] A logged-in user is able to send themselves a friend request. This should not be the expected app's behaviour. Give this a quick fix.
- [x] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from Pathwright)
- [x] After sending a request from the single user page, the send request button is still visible and a request can be sent numerous times, Remove the button and clear up this issue.
- [x] Accepting a friend request is currently not successful. Give this issue a quick look into it.
- [x] Check the inline comments below for more sections to improve

#

### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

<hr>

Mention @mke2111 in this thread and I'll be glad to help. 😄 
### Best regards
Slack 🆘  Roy Mukuye
Twitter 🐦  [Roy Mukuye](https://twitter.com/Roymkenya)
LinkedIn 🔗  [Roy Mukuye](https://www.linkedin.com/in/roy-mukuye/)

File: app/models/user.rb [hunk 1]
```diff
@@ -9,4 +9,39 @@ class User < ApplicationRecord
   has_many :posts
   has_many :comments, dependent: :destroy
   has_many :likes, dependent: :destroy
+  has_many :friendships # sent requests
+  has_many :inverse_friendships, class_name: 'Friendship', foreign_key: 'friend_id' # received requests
+
+  has_many :pending_friendships, -> { where(confirmed: false) }, class_name: 'Friendship', foreign_key: 'user_id'
+  has_many :pending_friends, through: :pending_friendships, source: :friend_id
```
Comment (mke2111) [inline app/models/user.rb]: - [ ] Remove these two lines of SQL queries. They'll come in handy in the next milestone. The ruby enumerables will do the job perfectly for now.

File: app/models/user.rb [hunk 2]
```diff
@@ -9,4 +9,39 @@ class User < ApplicationRecord
...
+
+  has_many :accepted_friendships, -> { where(confirmed: true) }, class_name: 'Friendship'
+  has_many :friends, through: :accepted_friendships, source: :friend_id
```
Comment (mke2111) [inline app/models/user.rb]: - [ ] Remove these two lines of SQL queries. They'll come in handy in the next milestone. The ruby enumerables will do the job perfectly for now.

File: app/models/user.rb [hunk 3]
```diff
@@ -9,4 +9,39 @@ class User < ApplicationRecord
...
+
+  has_many :inverse_friendships, -> { where(confirmed: false) }, class_name: 'Friendship', foreign_key: 'friend_id'
+  has_many :friend_requests, through: :inverse_friendships, source: :user_id
```
Comment (mke2111) [inline app/models/user.rb]: - [ ] Remove these two lines of SQL queries. They'll come in handy in the next milestone. The ruby enumerables will do the job perfectly for now.

**Review by elisha2kyakpo1 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️
Hey, @Leon-Mbegera you have done really good, this is impressive. From the last reviewer's comments, I have seen that some changes are not being addressed.

## To highlight

- You have added tests, which is good.
- The pr has descriptive comments.
- A user can send a request for friendship.

1. The app runs just fine
However, there are a few sections to improve before your project is approved. 🔴

REQUIRED
 

- [x] Posts timeline includes posts written by logged-in user and their friends. Currently, every logged-in user can see a post from every user, even from users who are not friends.
- [x]  There is a repeat of friend requests, which in turn created a duplicate friendship. See from the image below:
repeat
repeat1

**Kindly make all the requested changes and submit another CRR. 💯**
Cheers 🍹 and Happy coding 💻

<hr>
In case you have a question, just mention **elisha2kyakpo1**, and I will respond.

**Review by SajjadAhmad14 — CHANGES_REQUESTED** [review]
# Changes Required :recycle:
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hello, you have been doing a tremendous job so far :ok_hand:
## To Highlight 🎉
- [x] Git flow is used properly.
- [x] README includes instructions to run the project locally.
## To Improve 🙏
- [x] User should be able to accept friendship requests to start a friendship. Please see this [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png).
- [x] User can reject a friendship request. Please see this [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png).
- [x] Posts timeline should be of posts written by logged-in user and their friends. Please see this [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png).
# Optional
- [ ] Views should be completely separate from app logic. You can move all view-related logic to app/helpers.
### Happy Coding ☺️
## If you have any questions you can comment here in this PR or can always connect with me on slack ( @Sajjad Ahmad ).

**Review by SajjadAhmad14 — CHANGES_REQUESTED** [review]
# Changes Required :recycle:
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hello, you have been doing a tremendous job so far :ok_hand:
## To Improve 🙏
- [x] User should be able to accept friendship requests to start a friendship. Please see this [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png).
- [x] User can reject a friendship request. Please see this [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png).
- [x] Posts timeline should be of posts written by logged-in user and their friends. Please see this [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png).
# Optional
- [ ] Views should be completely separate from app logic. You can move all view-related logic to app/helpers.
### Happy Coding ☺️
## If you have any questions you can comment here in this PR or can always connect with me on slack ( @Sajjad Ahmad ).

**Review by SajjadAhmad14 — APPROVED** [review]
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:
## Congratulations 🎉 Thank you for implementing changes suggested previously 🙏
### Happy Coding ☺️
## If you have any questions you can comment here in this PR or can always connect with me on slack ( @Sajjad Ahmad ).

**Pull request merged**